app/policies/namespace_policy.rb

```diff
@@ -40,20 +40,25 @@ def index?
     user.admin? || namespace.team.users.exists?(user.id)
   end
 
+  def create?
+    raise Pundit::NotAuthorizedError, "must be logged in" unless user
+    (APP_CONFIG.enabled?("user_permission.modify_namespace") || user.admin?) && push?
+  end
+
+  alias update? create?
   alias all? push?
-  alias create? push?
-  alias update? push?
 
   def change_visibility?
     raise Pundit::NotAuthorizedError, "must be logged in" unless user
-    (namespace.global? && user.admin?) || \
-      (!namespace.global? && (user.admin? || namespace.team.owners.exists?(user.id)))
+    user.admin? || (APP_CONFIG.enabled?("user_permission.change_visibility") &&
+                    !namespace.global? && namespace.team.owners.exists?(user.id))
```

Comment:
The message "must be logged in" is only shown if a user is not logged in. In case of prohibition by APP_CONFIG, the user should just get a 401. The Ruby code might look a bit confusing at first.